File: game-engine/src/test/java/org/luxons/sevenwonders/game/resources/BestPriceCalculatorTest.java
package org.luxons.sevenwonders.game.resources;

import java.util.Arrays;

import org.junit.Test;
import org.luxons.sevenwonders.game.api.Table;
import org.luxons.sevenwonders.game.boards.Board;
import org.luxons.sevenwonders.game.test.TestUtilsKt;

import static org.junit.Assert.assertEquals;
import static org.luxons.sevenwonders.game.resources.Provider.LEFT_PLAYER;
import static org.luxons.sevenwonders.game.resources.Provider.RIGHT_PLAYER;
import static org.luxons.sevenwonders.game.resources.ResourceType.CLAY;
import static org.luxons.sevenwonders.game.resources.ResourceType.GLASS;
import static org.luxons.sevenwonders.game.resources.ResourceType.ORE;
import static org.luxons.sevenwonders.game.resources.ResourceType.STONE;
import static org.luxons.sevenwonders.game.resources.ResourceType.WOOD;

public class BestPriceCalculatorTest {

    @Test
    public void bestPrice_0forEmptyResources() {
        Table table = TestUtilsKt.testTable(3);
        Resources resources = new Resources();
        assertEquals(0, BestPriceCalculator.bestPrice(resources, table, 0));
        assertEquals(new ResourceTransactions(), BestPriceCalculator.bestSolution(resources, table, 0));
    }

    @Test
    public void bestPrice_fixedResources_defaultCost() {
        Board left = TestUtilsKt.testBoard(STONE);
        Board main = TestUtilsKt.testBoard(STONE);
        Board right = TestUtilsKt.testBoard(WOOD);
        Table table = new Table(Arrays.asList(main, right, left));

        Resources resources = TestUtilsKt.createResources(STONE, STONE);
        assertEquals(2, BestPriceCalculator.bestPrice(resources, table, 0));
        assertEquals(4, BestPriceCalculator.bestPrice(resources, table, 1));
        assertEquals(2, BestPriceCalculator.bestPrice(resources, table, 2));

        ResourceTransaction stoneLeftSingle = TestUtilsKt.createTransaction(LEFT_PLAYER, STONE);
        ResourceTransaction stoneRightSingle = TestUtilsKt.createTransaction(RIGHT_PLAYER, STONE);

        ResourceTransactions stoneLeft = TestUtilsKt.createTransactions(stoneLeftSingle);
        ResourceTransactions stoneRight = TestUtilsKt.createTransactions(stoneRightSingle);
        ResourceTransactions stoneLeftAndRight = TestUtilsKt.createTransactions(stoneLeftSingle, stoneRightSingle);

        assertEquals(stoneLeft, BestPriceCalculator.bestSolution(resources, table, 0));
        assertEquals(stoneLeftAndRight, BestPriceCalculator.bestSolution(resources, table, 1));
        assertEquals(stoneRight, BestPriceCalculator.bestSolution(resources, table, 2));
    }

    @Test
    public void bestPrice_fixedResources_overridenCost() {
        Board main = TestUtilsKt.testBoard(STONE);
        main.getTradingRules().setCost(WOOD, RIGHT_PLAYER, 1);

        Board left = TestUtilsKt.testBoard(WOOD);
        Board right = TestUtilsKt.testBoard(WOOD);
        Board opposite = TestUtilsKt.testBoard(GLASS);
        Table table = new Table(Arrays.asList(main, right, opposite, left));

        Resources resources = TestUtilsKt.createResources(WOOD);
        assertEquals(1, BestPriceCalculator.bestPrice(resources, table, 0));
        assertEquals(0, BestPriceCalculator.bestPrice(resources, table, 1));
        assertEquals(2, BestPriceCalculator.bestPrice(resources, table, 2));
        assertEquals(0, BestPriceCalculator.bestPrice(resources, table, 3));

        ResourceTransactions woodLeft = TestUtilsKt.createTransactions(LEFT_PLAYER, WOOD);
        ResourceTransactions woodRight = TestUtilsKt.createTransactions(RIGHT_PLAYER, WOOD);
        assertEquals(woodRight, BestPriceCalculator.bestSolution(resources, table, 0));
        assertEquals(new ResourceTransactions(), BestPriceCalculator.bestSolution(resources, table, 1));
        assertEquals(woodLeft, BestPriceCalculator.bestSolution(resources, table, 2));
        assertEquals(new ResourceTransactions(), BestPriceCalculator.bestSolution(resources, table, 3));
    }

    @Test
    public void bestPrice_mixedResources_overridenCost() {
        Board left = TestUtilsKt.testBoard(WOOD);

        Board main = TestUtilsKt.testBoard(STONE);
        main.getTradingRules().setCost(WOOD, RIGHT_PLAYER, 1);

        Board right = TestUtilsKt.testBoard(ORE);
        right.getProduction().addChoice(WOOD, CLAY);
        right.getPublicProduction().addChoice(WOOD, CLAY);

        Table table = new Table(Arrays.asList(main, right, left));

        Resources resources = TestUtilsKt.createResources(WOOD);
        assertEquals(1, BestPriceCalculator.bestPrice(resources, table, 0));
        assertEquals(0, BestPriceCalculator.bestPrice(resources, table, 1));
        assertEquals(0, BestPriceCalculator.bestPrice(resources, table, 2));

        ResourceTransactions woodRight = TestUtilsKt.createTransactions(RIGHT_PLAYER, WOOD);

        assertEquals(woodRight, BestPriceCalculator.bestSolution(resources, table, 0));
        assertEquals(new ResourceTransactions(), BestPriceCalculator.bestSolution(resources, table, 1));
        assertEquals(new ResourceTransactions(), BestPriceCalculator.bestSolution(resources, table, 2));
    }

    @Test
    public void bestPrice_chooseCheapest() {
        Board left = TestUtilsKt.testBoard(WOOD);

        Board main = TestUtilsKt.testBoard(WOOD);
        main.getProduction().addChoice(CLAY, ORE);
        main.getTradingRules().setCost(CLAY, RIGHT_PLAYER, 1);

        Board right = TestUtilsKt.testBoard(WOOD);
        right.getProduction().addFixedResource(ORE, 1);
        right.getProduction().addFixedResource(CLAY, 1);
        right.getPublicProduction().addFixedResource(ORE, 1);
        right.getPublicProduction().addFixedResource(CLAY, 1);

        Table table = new Table(Arrays.asList(main, right, left));

        Resources resources = TestUtilsKt.createResources(ORE, CLAY);
        assertEquals(1, BestPriceCalculator.bestPrice(resources, table, 0));
        assertEquals(0, BestPriceCalculator.bestPrice(resources, table, 1));
        assertEquals(4, BestPriceCalculator.bestPrice(resources, table, 2));

        ResourceTransactions oreAndClayLeft = TestUtilsKt.createTransactions(LEFT_PLAYER, ORE, CLAY);
        ResourceTransactions clayRight = TestUtilsKt.createTransactions(RIGHT_PLAYER, CLAY);
        assertEquals(clayRight, BestPriceCalculator.bestSolution(resources, table, 0));
        assertEquals(new ResourceTransactions(), BestPriceCalculator.bestSolution(resources, table, 1));
        assertEquals(oreAndClayLeft, BestPriceCalculator.bestSolution(resources, table, 2));
    }
}

File: game-engine/src/test/kotlin/org/luxons/sevenwonders/game/resources/BestPriceCalculatorTest.kt
package org.luxons.sevenwonders.game.resources

import org.junit.Assert.assertEquals
import org.junit.Test
import org.luxons.sevenwonders.game.SimplePlayer
import org.luxons.sevenwonders.game.api.Table
import org.luxons.sevenwonders.game.resources.Provider.LEFT_PLAYER
import org.luxons.sevenwonders.game.resources.Provider.RIGHT_PLAYER
import org.luxons.sevenwonders.game.resources.ResourceType.CLAY
import org.luxons.sevenwonders.game.resources.ResourceType.GLASS
import org.luxons.sevenwonders.game.resources.ResourceType.ORE
import org.luxons.sevenwonders.game.resources.ResourceType.STONE
import org.luxons.sevenwonders.game.resources.ResourceType.WOOD
import org.luxons.sevenwonders.game.test.createTransaction
import org.luxons.sevenwonders.game.test.createTransactions
import org.luxons.sevenwonders.game.test.testBoard
import org.luxons.sevenwonders.game.test.testTable
import java.util.Arrays

class BestPriceCalculatorTest {

    private fun solutions(price: Int, vararg resourceTransactions: ResourceTransactions) =
        TransactionPlan(price, setOf(*resourceTransactions))

    @Test
    fun bestPrice_0forEmptyResources() {
        val table = testTable(3)
        val player0 = SimplePlayer(0, table)
        val emptyResources = emptyResources()
        val emptyTransactions = noTransactions()
        assertEquals(solutions(0, emptyTransactions), bestSolution(emptyResources, player0))
    }

    @Test
    fun bestPrice_fixedResources_defaultCost() {
        val left = testBoard(STONE)
        val main = testBoard(STONE)
        val right = testBoard(WOOD)
        val table = Table(Arrays.asList(main, right, left))

        val player0 = SimplePlayer(0, table)
        val player1 = SimplePlayer(1, table)
        val player2 = SimplePlayer(2, table)

        val resources = resourcesOf(STONE, STONE)

        val stoneLeftSingle = createTransaction(LEFT_PLAYER, STONE)
        val stoneRightSingle = createTransaction(RIGHT_PLAYER, STONE)

        val stoneLeft = createTransactions(stoneLeftSingle)
        val stoneRight = createTransactions(stoneRightSingle)
        val stoneLeftAndRight = createTransactions(stoneLeftSingle, stoneRightSingle)

        assertEquals(solutions(2, stoneLeft), bestSolution(resources, player0))
        assertEquals(solutions(4, stoneLeftAndRight), bestSolution(resources, player1))
        assertEquals(solutions(2, stoneRight), bestSolution(resources, player2))
    }

    @Test
    fun bestPrice_fixedResources_overridenCost() {
        val main = testBoard(STONE)
        main.tradingRules.setCost(WOOD, RIGHT_PLAYER, 1)

        val left = testBoard(WOOD)
        val right = testBoard(WOOD)
        val opposite = testBoard(GLASS)
        val table = Table(Arrays.asList(main, right, opposite, left))

        val player0 = SimplePlayer(0, table)
        val player1 = SimplePlayer(1, table)
        val player2 = SimplePlayer(2, table)
        val player3 = SimplePlayer(3, table)

        val resources = resourcesOf(WOOD)

        val woodLeft = createTransactions(LEFT_PLAYER, WOOD)
        val woodRight = createTransactions(RIGHT_PLAYER, WOOD)

        assertEquals(solutions(1, woodRight), bestSolution(resources, player0))
        assertEquals(solutions(0, noTransactions()), bestSolution(resources, player1))
        assertEquals(solutions(2, woodLeft, woodRight), bestSolution(resources, player2))
        assertEquals(solutions(0, noTransactions()), bestSolution(resources, player3))
    }

    @Test
    fun bestPrice_mixedResources_overridenCost() {
        val left = testBoard(WOOD)

        val main = testBoard(STONE)
        main.tradingRules.setCost(WOOD, RIGHT_PLAYER, 1)

        val right = testBoard(ORE)
        right.production.addChoice(WOOD, CLAY)
        right.publicProduction.addChoice(WOOD, CLAY)

        val table = Table(Arrays.asList(main, right, left))

        val player0 = SimplePlayer(0, table)
        val player1 = SimplePlayer(1, table)
        val player2 = SimplePlayer(2, table)

        val resources = resourcesOf(WOOD)
        val woodRight = createTransactions(RIGHT_PLAYER, WOOD)

        assertEquals(solutions(1, woodRight), bestSolution(resources, player0))
        assertEquals(solutions(0, noTransactions()), bestSolution(resources, player1))
        assertEquals(solutions(0, noTransactions()), bestSolution(resources, player2))
    }

    @Test
    fun bestPrice_chooseCheapest() {
        val left = testBoard(WOOD)

        val main = testBoard(WOOD)
        main.production.addChoice(CLAY, ORE)
        main.tradingRules.setCost(CLAY, RIGHT_PLAYER, 1)

        val right = testBoard(WOOD)
        right.production.addFixedResource(ORE, 1)
        right.production.addFixedResource(CLAY, 1)
        right.publicProduction.addFixedResource(ORE, 1)
        right.publicProduction.addFixedResource(CLAY, 1)

        val table = Table(Arrays.asList(main, right, left))

        val player0 = SimplePlayer(0, table)
        val player1 = SimplePlayer(1, table)
        val player2 = SimplePlayer(2, table)

        val resources = resourcesOf(ORE, CLAY)
        val oreAndClayLeft = createTransactions(LEFT_PLAYER, ORE, CLAY)
        val clayRight = createTransactions(RIGHT_PLAYER, CLAY)

        assertEquals(solutions(1, clayRight), bestSolution(resources, player0))
        assertEquals(solutions(0, noTransactions()), bestSolution(resources, player1))
        assertEquals(solutions(4, oreAndClayLeft), bestSolution(resources, player2))
    }
}
